py3status/modules/imap.py

```diff
@@ -65,61 +73,163 @@ class Meta:
             raise ValueError("Unknown security protocol")
 
-    def check_mail(self):
-        mail_count = self._get_mail_count()
-
-        response = {'cached_until': self.py3.time_in(self.cache_timeout)}
+        self.idle_thread = Thread(target=self._get_mail_count, daemon=True)
+        if self.use_idle:
+            self.idle_thread.start()
 
-        if mail_count is None:
+    def check_mail(self):
+        # I -- acquire mail_count
+        if self.use_idle is not False:
+            if not self.idle_thread.is_alive():
+                sleep(5)  # rate-limit thread-restarting (when network is offline)
+                self.idle_thread = Thread(target=self._get_mail_count, daemon=True)
+                self.idle_thread.start()
+            response = {'cached_until': self.py3.CACHE_FOREVER}
+        else:
+            self._get_mail_count()
+            response = {'cached_until': self.py3.time_in(self.cache_timeout)}
+        if self.mail_error is not None:
+            self.py3.log(self.mail_error, level=self.py3.LOG_ERROR)
+            self.py3.error(self.mail_error)
+
+        # II -- format response
+        if self.mail_count is None:
             response['color'] = self.py3.COLOR_BAD,
             response['full_text'] = self.py3.safe_format(
                 self.format, {'unseen': STRING_UNAVAILABLE})
-        elif mail_count > 0:
+        elif self.mail_count > 0:
             response['color'] = self.py3.COLOR_NEW_MAIL or self.py3.COLOR_GOOD
             if self.allow_urgent:
                 response['urgent'] = True
 
-        if mail_count == 0 and self.hide_if_zero:
+        if self.mail_count == 0 and self.hide_if_zero:
             response['full_text'] = ''
         else:
-            response['full_text'] = self.py3.safe_format(self.format, {'unseen': mail_count})
+            response['full_text'] = self.py3.safe_format(self.format, {'unseen': self.mail_count})
 
         return response
 
+    def _check_if_idle(self, connection):
+        supports_idle = 'IDLE' in connection.capabilities
+
+        if self.use_idle is None:
+            self.use_idle = supports_idle
+            self.py3.log("Will use {}".format('idling' if self.use_idle else 'polling'))
+        elif self.use_idle and not supports_idle:
+            self.py3.error("Server does not support IDLE")
+
     def _connection_ssl(self):
         connection = imaplib.IMAP4_SSL(self.server, self.port)
+        self._check_if_idle(connection)
         return connection
 
     def _connection_starttls(self):
         connection = imaplib.IMAP4(self.server, self.port)
         connection.starttls(create_default_context())
+        self._check_if_idle(connection)
         return connection
 
-    def _get_mail_count(self):
+    def _connect(self):
+        if self.security == "ssl":
+            self.connection = self._connection_ssl()
+        elif self.security == "starttls":
+            self.connection = self._connection_starttls()
```

Comment:
(Q) Is `self.connection = self._connection_*()` something we can put in `post_config_hook`? To see if we can skip connection object + check-if-idle repeatedly on every update (polling).

Reply:
connecting takes some time; so if we put that inside `_post_config_hook`, none of the modules load until we are connected. (i did do that a few commits back, but changed it because of this)

Reply:
Hmm. Check out `coin_market`, `github`, or `gcal`. There might be more... Those modules uses `first_run` to make sure they all return `''` right away before connecting... so all modules can load immediately. Sounds like this module could use it too.

Reply:
I don't see the value in returning immediately if we don't have any data to display. (also, the original imap module didn't do that either)

Reply:
>connecting takes some time; so if we put that inside _post_config_hook, none of the modules load until we are connected.

Just to be clear, returning immediately on first run will fix this problem even if we don't have anything to show. That's the gist or the value.

Indeed, original `imap` does not do that either. Maybe it could be something for us to look into... Does `imap` cause other modules to wait on startup and/or offline? Et cetera. Lot of modules/things to check. I recently fixed #1156 (offline bug) too.

I know `gcal` takes a whole minute or two before it finally throws an exception... so I don't do that in `post_config_hook` anymore and to use `self.first_run`.

Really, there is lot to do with all modules. This `imap` also need some formatting love too... to deprecate `hide_if_zero` in favor of `\?not_zero {count}` and other things.